File: Go/Day05/Day05.go
//go:build d05
// +build d05

package Day05

import (
	AH "AoC2024/adventhelper"
	"strconv"
	"strings"
)

type Rule struct {
	lhs, rhs int
}

func parseRulesToBook(s string) []Rule {
	rules := []Rule{}
	ss := strings.Split(s, "#")

	for _, r := range ss {
		ps := strings.Split(r, "|")

		l, _ := strconv.Atoi(ps[0])
		r, _ := strconv.Atoi(ps[1])

		rules = append(rules, Rule{l, r})
	}

	return rules
}

func sortBook(book []int, rules []Rule) ([]int, int) {
	counter := 0
	// bubble sort book into order
	for iters := 0; iters < len(book); iters++ {
		for i := 0; i < len(book)-1; i++ {
			for j := i + 1; j < len(book); j++ {
				for _, r := range rules {
					if (r.lhs == book[j]) && (r.rhs == book[i]) {
						temp := book[i]
						book[i] = book[j]
						book[j] = temp
						counter++
					}
				}
			}
		}
	}

	return book, counter
}

func parseBooks(s string) (bs [][]int) {
	ss := strings.Split(s, "#")
	for _, r := range ss {
		b := []int{}
		ps := strings.Split(r, ",")
		for _, p := range ps {
			pg, _ := strconv.Atoi(p)
			b = append(b, pg)
		}
		bs = append(bs, b)
	}

	return
}

func Run() {
	g, _ := AH.ParseLineGroups("../input/input05.txt", "#")
	rules := parseRulesToBook(g[0])
	books := parseBooks(g[1])

	part1, part2 := 0, 0

	for _, book := range books {
		ordered, count := sortBook(book, rules)
		if count == 0 {
			part1 += ordered[len(book)/2]
		} else {
			part2 += ordered[len(book)/2]
		}
	}

	AH.PrintSoln(5, part1, part2)

	return
}

File: C++/Day05.cpp
#ifdef DAY05

#include "AH.h"

namespace Day05
{

	typedef std::pair<int,int> Rule;
	
	std::vector<Rule> parseRules(const std::string s)
	{
		std::vector <Rule> rs;
		const auto ss = AH::Split(s, '#');

		for (auto r : ss) {
			const auto ps = AH::Split(r, '|');
			auto lhs = std::stoi(ps[0]);
			auto rhs = std::stoi(ps[1]);
			rs.emplace_back(lhs, rhs);
		}

		return rs;
	}

	std::vector<std::vector<int>> parseBooks(const std::string s)
	{
		std::vector <std::vector<int>> bs;
		const auto ss = AH::Split(s, '#');
		for (auto r : ss) {
			std::vector<int> b;
			const auto ps = AH::Split(r, ',');
			for (auto p : ps) {
				auto pg = std::stoi(p);
				b.emplace_back(pg);
			}
			bs.emplace_back(b);
		}

		return bs;
	}

	std::pair<std::vector<int>, int> sortBook (
		const std::vector<Rule> rs,
		const std::vector<int> b
	)
	{
		std::vector<int>new_book = b;
		int counter = 0;
		// do n bubble sorts cause i am an scrub
		for (size_t iters = 0; iters < b.size(); ++iters) {
			for (auto  p1 = new_book.begin(); p1 != new_book.end() - 1; p1++) {
				for (auto p2 = p1 + 1; p2 != new_book.end(); p2++) {

					// does this breach a rule?
					for (auto r : rs) {
						if ((r.first == *p2) && (r.second == *p1)) {
							std::iter_swap(p1, p2);
							counter++;
						}
					}
				}
			}		   
		}

		return std::make_pair(new_book, counter);
	}

	int Run(const std::string& filename)
	{
		const auto is = AH::ReadTextFile(filename);
		const auto ss = AH::ParseLineGroups(is ,'#');
		const auto rules = parseRules(ss[0]);
		const auto books = parseBooks(ss[1]);

		int part1 = 0;
		int part2 = 0;
		for (auto book : books) {
			auto [ordered, count] = sortBook(rules, book);
			if (count == 0) {
				part1 += ordered[book.size() / 2];
			} else {
				part2 += ordered[book.size() / 2];
			}
		}

		AH::PrintSoln(5, part1, part2);

		return 0;
	}

}

#endif
